### ISSUE TITLE
tsp, support multiple content types

### ISSUE THREAD

**Issue by weidongxu-microsoft** [Azure/autorest.java#1958]
epic: https://github.com/Azure/cadl-azure/issues/1943

**Comment by weidongxu-microsoft** [comment]
pending on Cadl source

**Comment by haolingdong-msft** [comment]
I wrote a draft design [doc](https://github.com/haolingdong-msft/working-docs/blob/main/cadl/multi-content-type.md) on it. I first list different user cases with the proposed generated SDK. Once we agree on what should SDK look like, I will start design on the codegen implementation part. Please take a look and let me know your thoughts.
/cc @srnagar @weidongxu-microsoft @XiaofeiCao

**Comment by weidongxu-microsoft** [comment]
### Case 1

For multi content-type without overload, inclined to only generate protocol API.
In swagger, there is no way to properly define what method signature it want to generate, hence it is kind of reasonable for modelerfour to make up a solution (likely sperate JSON / text / binary types).

Given now Cadl provide the "overload", we'd expect user use that.

Multi content-type is not a common case, I don't think adding a few "overload"s a big effort on service (and it applies to all SDKs, not only Java).

### Case 2

I think SDK should not use `byte[]` in customer-facing API. It should still be `upload(BinaryData data, ContentType contentType)`.

Agree that we'd need some consensus (maybe across language) on what "overload" is allowed.
Probably not want to see this https://gist.github.com/msyyc/87a79b917deec25639f9f10d20589e55?permalink_comment_id=4472749#gistcomment-4472749

**Comment by haolingdong-msft** [comment]
Metting notes at 2023-02-22:

Case 1:

This case is not valid. This is agreed by all languages' archs. We will have lint rule later to prevent the definition which defines both union types in body and union types in header. User should define overload if they define body type as union and header type as union.
Before that, we can throw error when finding this case in emitter or use option1 to generate protocol method only.

Case 2:
- `uploadStringOrResource` is not valid, similar to case1, should prevent in the linter.
- We agree on the overload methods should be the same name.
- For the body types that do not define overload method, we will not generate convenient methods for them, in this case, `Resource` type with `application/json`. Later we can also consider a lint rule to make sure all the types in the union body have defined overload operations.
- For overload with same signature case, we will generate two overload methods with compilation error.
- We should use `BinaryData` instead of `byte[]`

Other cases:

- Overload on response: previous agreement is to use different name for overload on response, but this makes overload names tricky.  Agreement in the meeting is that we will not handle overload on response, should suggest customer to use shared route.
- We will not support `@internal` for now.

Follow up items for @srnagar:
- Check with other languages' archs about overload on response. We will not support overload on response, we should suggest customer to use shared route when response type is union.
- Check with python whey they consider `@internal` case.

**Comment by haolingdong-msft** [comment]
Meeting notes in 2020-03-01:

* overload on return type: we will support it with overload as much as we can, which means if input parameters are different, we will generate overload methods with the same name, if input parameters are the same, then we will overwrite the name in client.cadl. (Previously we want to recommand to use shared route in this case, but overload should be better when the nature of the operations are the same, e.g. all the operation means upload. Shared route will be used in other scenarios.)
* we can start the implementation for overload approach now.
* @internal: we will make it as low priority.

**Comment by haolingdong-msft** [comment]
Meeting notes in 2023-3-9:

- multiple response type, we will suggest service use shared route
- `@body` issue: if we make it implicit, we will get `Model` type from compiler and we need to handle differently for different content types.
- Archs suggest remove `@overload` decorator and the base operation, but if we want to have the base operation for implementation conveniency, we can add it in client.tsp and make it `@internal`.

**Comment by haolingdong-msft** [comment]
Meeting notes in 2023-3-23:
* implicit @body should only be applied to application/json content type, and it means a property of an unnamed model. If it is used in other content types, it should be invalid.

**Comment by haolingdong-msft** [comment]
Meeting notes in 2023-4-7

- We will always recommend users to use shared route to define multiple content types, but nothing prevents user to define multiple content types using overload, so if emitter meets this case, there are two solutions 1. throw error, but this requries linter rules and all languages have the same error. 2. emitter handles it and generate something, for Java, we decide to generate protocol methods only in cases where overload is defined with union types.
- [Java only] `byte[]` should always be represented as `BinaryData` in convenient methods no matter it's in multiple content type definition or not.
- [Java only] For shared route, we will basically generate protocol methods and convenient methods for each operations defined. But if the user overwrite the operation name in client.tsp to make the operations as the same name, we should have logic to handle this, we should only generate one protocol method and multiple convenient methods in this case.

**Comment by haolingdong-msft** [comment]
Meeting notes on 5-10

* Update from @srnagar  that whenever we see union, we will generate protocol method only, no matter it's union in body, parameter, response

**Comment by weidongxu-microsoft** [comment]
Whatever you do, make sure the "condition" is in a single place. So that it can be modified easily.

I may not agree with above, e.g. I don't think API with `name: string = "default" | "name"` be a multiple content-type. But apparently `content-type: "application/json" | "text"` is.

**Comment by haolingdong-msft** [comment]
`name: string = "default" | "name"` is not multiple content types, this is union, the logic is for union because we don't want the parttern we use currently to process union using base type and sub types. @srnagar Please correct me if I understand wrongly.
But I do think string union like `name: string = "default" | "name"` is a special case, which we will use enum, not base type and sub types as we do for normal unions, so in this case we can generate convenient methods. 

> Whatever you do, make sure the "condition" is in a single place. So that it can be modified easily.
> 
> I may not agree with above, e.g. I don't think API with `name: string = "default" | "name"` be a multiple content-type. But apparently `content-type: "application/json" | "text"` is.

**Comment by haolingdong-msft** [comment]
Meeting minutes for 2023-5-24 sync up:
- we will only generate protocol methods for multiple content types
- we will need to generate convenient method for `@overload` methods if it is not multiple content type (this part will not be included in the first pr)
